src/utils/helpers/DetailsDrawHelpers.ts

```diff
@@ -0,0 +1,223 @@
+    const containerWidth = width * 4;
+    const containerHeight = height * 4;
+    const container = selectDetailsContainerDiv()
+        .style('display', 'none')
+        .style('opacity', 0)
+        .append('svg')
+        .attr('id', 'details-view-container')
+        .attr('x', 0)
+        .attr('y', 0)
+        .attr('width', containerWidth)
+        .attr('height', containerHeight)
+        .attr('viewBox', `0 0 ${width} ${height}`)
+        .attr('preserveAspectRatio', 'xMidYMid meet');
+    container
+        .append('rect')
+        .attr('width', width)
+        .attr('height', height)
+        .attr('fill', ElementColors.DETAILS_BACKGROUND);
+    container
+        .append('g')
+        .attr('id', 'details-exit-button')
+        .attr('cursor', 'pointer')
+        .append('rect')
+        .attr('transform', 'translate(5,5)')
+        .attr('width', 10)
+        .attr('height', 10)
+        .attr('fill', ElementColors.DETAILS_BACKGROUND);
+    selectDetailsExitButtonWrapper()
+        .append('path')
+        .attr('transform', 'translate(10,10) rotate(45)')
+        .attr(
+            'd',
+            symbol()
+                .type(symbolCross)
+                .size(20)
+        );
+    createZoom(container, Selectors.ZOOM_DETAILS);
+    container.call(
+        zoom<any, any>().on('zoom', changeZoom(Selectors.ZOOM_DETAILS)).transform,
+        // rough center screen on diagram's root node
+        zoomIdentity.translate(-width / 5.3, -height / 2.65)
+    );
+}
+
+interface TreeStructure {
+    name: string;
+    children?: TreeStructure[];
+}
+
+type TreeNodeWithVisitedFlag = TreeNode & { isVisited?: boolean };
+
+function mapNodeToTreeStructure(node: TreeNodeWithVisitedFlag, links: (keyof TreeNode & 'consumers') | 'providers'): TreeStructure {
+    node.isVisited = true;
+    const UnvisitedLinks = node[links].filter((node: TreeNodeWithVisitedFlag) => !node.isVisited && node[links]);
+    const children = UnvisitedLinks.length > 0 ? UnvisitedLinks.map(nestedNode => mapNodeToTreeStructure(nestedNode, links)) : [];
+    node.isVisited = undefined;
+    return { name: node.name, children };
+}
+
+const VERTICAL_DISTANCE_BETWEEN_NODES = 40;
+const HORIZONTAL_DISTANCE_BETWEEN_NODES = 300;
+
+const NODE_NEIGHBOURS_SEPARATION_MULTIPLIER = 1;
+const NODE_NORMAL_SEPARATION_MULTIPLIER = 4;
+
+function createTree(data: TreeStructure) {
+    const node = hierarchy(data);
+    return tree<TreeStructure>()
+        .nodeSize([VERTICAL_DISTANCE_BETWEEN_NODES, HORIZONTAL_DISTANCE_BETWEEN_NODES])
+        .separation((node1, node2) =>
+            node1.parent === node2.parent ? NODE_NEIGHBOURS_SEPARATION_MULTIPLIER : NODE_NORMAL_SEPARATION_MULTIPLIER
+        )(node);
+}
+
+function getRootYPosition(data: HierarchyPointNode<TreeStructure>) {
+    let x0 = Infinity;
+    let x1 = -x0;
+    data.each(node => {
+        if (node.x > x1) x1 = node.x;
+        if (node.x < x0) x0 = node.x;
+    });
```

Comment:
Are you looking for the smallest x ? 

Math.min(...data.map(node => node.x))

Reply:
data is a d3 object - it is easier to use native d3 function for iterating through it than try to map data to array and then operate on it